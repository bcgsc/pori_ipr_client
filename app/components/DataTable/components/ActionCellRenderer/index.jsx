import React, { useState, useEffect } from 'react';
import {
  IconButton,
  Menu,
  MenuItem,
} from '@mui/material';
import {
  Edit,
  Photo,
  LibraryBooks,
  OpenInNew,
  Delete,
} from '@mui/icons-material';
import DetailDialog from '../DetailDialog';
import SvgViewer from '../SvgViewer';
import ImageViewer from '../ImageViewer';

import './index.scss';

/**
 * @param {object} params params
 * @param {string} params.value display text
 * @param {string} props.link target link
 * @param {function} props.onEdit handler which is trigger on the user clicking the edit icon for this row
 * @return {*} JSX
 */
function ActionCellRenderer(params) {
  const {
    data,
    context: {
      canEdit,
      canViewDetails,
      canDelete,
    },
    columnApi,
    onEdit,
    onDelete,
  } = params;

  const [showDetailDialog, setShowDetailDialog] = useState(false);
  const [showSvgViewer, setShowSvgViewer] = useState(false);
  const [showImageViewer, setShowImageViewer] = useState(false);
  const [columnMapping, setColumnMapping] = useState({});
  const [anchorEl, setAnchorEl] = useState(null);

  useEffect(() => {
    if (showDetailDialog) {
      setColumnMapping(
        columnApi.getAllColumns().reduce((accumulator, current) => {
          accumulator[current.colId] = columnApi.getDisplayNameForColumn(current);
          return accumulator;
        }, {}),
      );
    }
  }, [columnApi, showDetailDialog]);

  const detailClick = () => {
    setShowDetailDialog(true);
  };

  const handleDetailClose = () => {
    setShowDetailDialog(false);
  };

  const handleSvgViewerClose = () => {
    setShowSvgViewer(false);
  };

  const handleImageViewerClose = () => {
    setShowImageViewer(false);
  };

  const handleMenuOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  return (
    <>
      {canViewDetails && (
        <IconButton
          size="small"
          aria-label="View Details"
          onClick={detailClick}
          title="View Details"
        >
          <LibraryBooks />
        </IconButton>
      )}
      {(!window._env_.IS_DEMO && data.kbStatementId && !Array.isArray(data.kbStatementId) && data.kbStatementId.match(/^#?-?\d+:-?\d+$/))
        ? (
          <IconButton
            size="small"
            aria-label="Open in GraphKB"
            title="Open in GraphKB"
            href={`${window._env_.GRAPHKB_URL}/view/Statement/${data.kbStatementId.replace('#', '')}`}
            target="_blank"
            rel="noreferrer noopener"
          >
            <OpenInNew />
          </IconButton>
        ) : null
      }
      {(!window._env_.IS_DEMO && data.kbStatementId && Array.isArray(data.kbStatementId) && data.kbStatementId.some(statement => statement.match(/^#?-?\d+:-?\d+$/)))
        ? (
          <>
            <IconButton
              size="small"
              aria-label="Open in GraphKB"
              title="Open in GraphKB"
              onClick={handleMenuOpen}
            >
              <OpenInNew />
            </IconButton>
            <Menu
              anchorEl={anchorEl}
              open={Boolean(anchorEl)}
              onClose={handleMenuClose}
            >
              {data.kbStatementId.filter(statement => statement.match(/^#?-?\d+:-?\d+$/)).map(statement => (
                <MenuItem
                  key={statement}
                  onClick={handleMenuClose}
                >
                  <a
                    className="action-cell-kb-statement__link"
                    href={`${window._env_.GRAPHKB_URL}/view/Statement/${statement.replace('#', '')}`}
                    target="_blank"
                    rel="noreferrer noopener"
                  >
                    {statement}
                  </a>
                </MenuItem>
              ))}
            </Menu>
          </>
        ) : null
      }
      {showDetailDialog && (
        <DetailDialog
          isOpen={showDetailDialog}
          selectedRow={data}
          onClose={handleDetailClose}
          columnMapping={columnMapping}
        />
      )}
      {canDelete && (
        <IconButton
          size="small"
          aria-label="Delete"
          onClick={() => onDelete(data.ident)}
          title="Delete"
        >
          <Delete />
        </IconButton>
      )}
      {canEdit && (
        <IconButton
          size="small"
          aria-label="Edit"
          onClick={() => onEdit(data)}
          title="Edit"
        >
          <Edit />
        </IconButton>
      )}
      {data.svg && (
        <IconButton
          size="small"
          aria-label="View Fusion Diagram"
          onClick={() => setShowSvgViewer(prevVal => !prevVal)}
          title="View Fusion Diagram"
        >
          <Photo />
        </IconButton>
      )}
      {showSvgViewer && (
        <SvgViewer
          isOpen={showSvgViewer}
          selectedRow={data}
          onClose={handleSvgViewerClose}
        />
      )}
      {data.image && (
        <IconButton
          size="small"
          aria-label="View Image"
          onClick={() => setShowImageViewer(prevVal => !prevVal)}
          title="View Image"
        >
          <Photo />
        </IconButton>
      )}
      {showImageViewer && (
        <ImageViewer
          isOpen={showImageViewer}
          selectedRow={data}
          onClose={handleImageViewerClose}
        />
      )}
    </>
  );
}

export default ActionCellRenderer;
