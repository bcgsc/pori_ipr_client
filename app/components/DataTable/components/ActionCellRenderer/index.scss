.action-cell-kb-statement {
  &__link {
    text-decoration: none;
    color: inherit;
  }
}