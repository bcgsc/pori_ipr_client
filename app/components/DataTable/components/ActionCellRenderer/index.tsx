import React, {
  useState, useEffect, useMemo, useCallback,
} from 'react';
import {
  IconButton,
  MenuItem,
} from '@mui/material';
import {
  Edit,
  Photo,
  LibraryBooks,
  OpenInNew,
  Delete,
} from '@mui/icons-material';
// eslint-disable-next-line import/no-extraneous-dependencies
import { ICellRendererParams } from '@ag-grid-community/core';

import DetailDialog from '../DetailDialog';
import SvgViewer from '../SvgViewer';
import ImageViewer from '../ImageViewer';

import './index.scss';

type ActionCellRendererProps = {
  // Finer Permission control on if row can be Edited, takes precedence over table canEdit via context of AgGridReact
  canEditRowData?: boolean;
  context?: {
    canEdit?: boolean;
    canViewDetails?: boolean;
    canDelete?: boolean;
  };
  onEdit?: (data) => void;
  onDelete?: (data: string) => void;
  displayMode?: 'tableCell' | 'menu';
} & Partial<ICellRendererParams>;

const WrapperComponent = ({
  displayMode = 'tableCell',
  children,
  onClick = () => {},
  ...rest
}: {
  displayMode: ActionCellRendererProps['displayMode'];
  children: React.ReactNode;
  onClick?: React.MouseEventHandler;
}) => {
  if (displayMode === 'menu') {
    return (
      <MenuItem onClick={onClick} {...rest}>
        {children}
      </MenuItem>
    );
  }

  // We don't deal with multiple nodes for now
  if (Array.isArray(children)) {
    // eslint-disable-next-line react/jsx-no-useless-fragment
    return <span {...rest}>{children}</span>;
  }

  if (React.isValidElement(children)) {
    const newProps = {
      onClick,
      ...rest,
    };
    return React.cloneElement(children, newProps);
  }
  return null;
};

const ActionCellRenderer = ({
  data,
  // Row-specific edit permissions
  canEditRowData,
  context: {
    canEdit: canEditTable,
    canViewDetails,
    canDelete,
  } = {},
  columnApi,
  onEdit,
  onDelete,
  displayMode = 'tableCell',
}: ActionCellRendererProps): JSX.Element => {
  const [showDetailDialog, setShowDetailDialog] = useState(false);
  const [showSvgViewer, setShowSvgViewer] = useState(false);
  const [showImageViewer, setShowImageViewer] = useState(false);
  const [columnMapping, setColumnMapping] = useState({});

  const canEdit = useMemo(() => {
    if (canEditRowData === undefined) {
      return canEditTable;
    }
    return canEditRowData;
  }, [canEditRowData, canEditTable]);

  useEffect(() => {
    if (showDetailDialog) {
      setColumnMapping(
        columnApi.getAllColumns().reduce((accumulator, current) => {
          accumulator[current.getColId()] = columnApi.getDisplayNameForColumn(current, null);
          return accumulator;
        }, {}),
      );
    }
  }, [columnApi, showDetailDialog]);

  const detailClick = () => {
    setShowDetailDialog(true);
  };

  const handleDetailClose = () => {
    setShowDetailDialog(false);
  };

  const handleSvgViewerClose = () => {
    setShowSvgViewer(false);
  };

  const handleImageViewerClose = () => {
    setShowImageViewer(false);
  };

  const handleSingleGraphKbLink = useCallback(() => {
    window.open(`${window._env_.GRAPHKB_URL}/view/Statement/${data.kbStatementId.replace('#', '')}`, '_blank');
  }, [data.kbStatementId]);

  const handleMultGraphKbLink = useCallback(() => {
    const sortedStatementIds = [...new Set(data.kbStatementId)].sort();
    const complexParam = btoa(JSON.stringify({ target: sortedStatementIds }));
    const searchParams = new URLSearchParams();
    searchParams.set('complex', complexParam);
    searchParams.set('@class', 'statement');
    const urlParams = searchParams.toString();
    window.open(`${window._env_.GRAPHKB_URL}/data/table?${urlParams}`, '_blank');
  }, [data.kbStatementId]);

  const handleDelete = useCallback(() => {
    onDelete(data.ident);
  }, [data.ident, onDelete]);

  const handleEdit = useCallback(() => {
    onEdit(data);
  }, [data, onEdit]);

  const hasImageData = Array.isArray(data.image) ? data.image.length > 0 : data.image;

  const openGraphKbButton = useMemo(() => {
    if (!window._env_.IS_DEMO && data.kbStatementId) {
      if (!Array.isArray(data.kbStatementId) && data.kbStatementId.match(/^#?-?\d+:-?\d+$/)) {
        return (
          <WrapperComponent
            data-testid="graphkb"
            onClick={handleSingleGraphKbLink}
            displayMode={displayMode}
          >
            {
              displayMode === 'tableCell' ? (
                <IconButton
                  aria-label="Open in GraphKB"
                  size="small"
                  title="Open in GraphKB"
                >
                  <OpenInNew />
                </IconButton>
              ) : 'Open in GraphKB'
            }
          </WrapperComponent>
        );
      }
      if (Array.isArray(data.kbStatementId) && data.kbStatementId.some((statement) => statement.match(/^#?-?\d+:-?\d+$/))) {
        return (
          <WrapperComponent onClick={handleMultGraphKbLink} displayMode={displayMode}>
            {displayMode === 'tableCell' ? (
              <IconButton
                size="small"
                aria-label="Open in GraphKB"
                title="Open in GraphKB"
              >
                <OpenInNew />
              </IconButton>
            ) : 'Open in GraphKB'}
          </WrapperComponent>
        );
      }
    }
    return null;
  }, [data.kbStatementId, displayMode, handleSingleGraphKbLink, handleMultGraphKbLink]);

  return (
    <>
      {canViewDetails && (
        <WrapperComponent data-testid="view-details" onClick={detailClick} displayMode={displayMode}>
          {displayMode === 'tableCell' ? (
            <IconButton
              aria-label="View Details"
              size="small"
              title="View Details"
            >
              <LibraryBooks />
            </IconButton>
          ) : 'View Details'}
        </WrapperComponent>
      )}
      {showDetailDialog && (
        <DetailDialog
          isOpen={showDetailDialog}
          selectedRow={data}
          onClose={handleDetailClose}
          columnMapping={columnMapping}
        />
      )}
      {openGraphKbButton}
      {canDelete && (
        <WrapperComponent data-testid="delete" onClick={handleDelete} displayMode={displayMode}>
          {displayMode === 'tableCell' ? (
            <IconButton
              aria-label="Delete"
              size="small"
              title="Delete"
            >
              <Delete />
            </IconButton>
          ) : 'Delete Row'}
        </WrapperComponent>
      )}
      {canEdit && (
        <WrapperComponent data-testid="edit" onClick={handleEdit} displayMode={displayMode}>
          {displayMode === 'tableCell' ? (
            <IconButton
              aria-label="Edit"
              size="small"
              title="Edit"
            >
              <Edit />
            </IconButton>
          ) : 'Edit Row'}
        </WrapperComponent>
      )}
      {data.svg && (
        <WrapperComponent data-testid="fusion" onClick={() => setShowSvgViewer((prevVal) => !prevVal)} displayMode={displayMode}>
          {displayMode === 'tableCell' ? (
            <IconButton
              aria-label="View Fusion Diagram"
              size="small"
              title="View Fusion Diagram"
            >
              <Photo />
            </IconButton>
          ) : 'View Fusion Diagram'}
        </WrapperComponent>
      )}
      {showSvgViewer && (
        <SvgViewer
          isOpen={showSvgViewer}
          selectedRow={data}
          onClose={handleSvgViewerClose}
        />
      )}
      {hasImageData && (
        <WrapperComponent data-testid="image" onClick={() => setShowImageViewer((prevVal) => !prevVal)} displayMode={displayMode}>
          {displayMode === 'tableCell' ? (
            <IconButton
              aria-label="View Image"
              size="small"
              title="View Image"
            >
              <Photo />
            </IconButton>
          ) : 'View Image'}
        </WrapperComponent>
      )}
      {showImageViewer && (
        <ImageViewer
          isOpen={showImageViewer}
          selectedRow={data}
          onClose={handleImageViewerClose}
        />
      )}
    </>
  );
};

export {
  ActionCellRendererProps,
  ActionCellRenderer,
};
export default ActionCellRenderer;
