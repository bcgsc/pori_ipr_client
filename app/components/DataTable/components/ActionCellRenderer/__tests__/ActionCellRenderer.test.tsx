import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ColumnApi } from '@ag-grid-community/core';

import ActionCellRenderer from '..';

// eslint-disable-next-line react/display-name
jest.mock('../../SvgViewer', () => (() => (<div role="presentation" />)));
// eslint-disable-next-line react/display-name
jest.mock('../../ImageViewer', () => (() => (<div role="presentation" />)));

const mockSvg = '<svg><g>this is an svg</g></svg>';
const mockImage = 'data:PNG;base64, image data here';

describe('ActionCellRenderer', () => {
  test('It matches the snapshot', () => {
    const { asFragment } = render(
      <ActionCellRenderer data={{}} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('It shows the detail icon when allowed', async () => {
    render(
      <ActionCellRenderer
        context={{ canViewDetails: true }}
        data={{}}
      />,
    );

    expect(await screen.findByTestId('view-details')).toBeInTheDocument();
  });

  test('It shows the detail dialog when the icon is clicked', async () => {
    const mockColumnApi = {
      getAllColumns: jest.fn(() => []),
    } as unknown as ColumnApi;

    render(
      <ActionCellRenderer
        columnApi={mockColumnApi}
        context={{ canViewDetails: true }}
        data={{}}
      />,
    );

    fireEvent.click(await screen.findByTestId('view-details'));
    expect(await screen.findAllByRole('presentation')).toHaveLength(2);
  });

  test('It links to GraphKB when a single entry exists', async () => {
    const openSpy = jest.spyOn(window, 'open').mockImplementation();
    render(
      <ActionCellRenderer
        data={{
          kbStatementId: '#155:863',
        }}
      />,
    );
    // Find the element and simulate a click event
    fireEvent.click(screen.getByTestId('graphkb'));

    // Assert that the window.open function was called with the expected URL and target
    expect(openSpy).toHaveBeenCalledWith(`${window._env_.GRAPHKB_URL}/view/Statement/155:863`, '_blank');
  });

  test('It shows a link to GraphKB when multiple entries exist', async () => {
    const statements = ['#155:863', '#120:109'];
    render(
      <ActionCellRenderer
        data={{
          kbStatementId: statements,
        }}
      />,
    );

    fireEvent.click(await screen.findByRole('button'));

    statements.forEach(async (statement) => {
      expect(await screen.findByText(statement))
        .toHaveAttribute('href', `${window._env_.GRAPHKB_URL}/view/Statement/${statement.replace('#', '')}`);
    });
  });

  test('It shows the delete icon when allowed', async () => {
    render(
      <ActionCellRenderer
        context={{ canDelete: true }}
        data={{}}
      />,
    );

    expect(await screen.findByTestId('delete')).toBeInTheDocument();
  });

  test('onDelete is called when the delete icon is clicked', async () => {
    const mockOnDelete = jest.fn();
    const mockIdent = 'ffa55bc-cc08852-b43a-b0fa';

    render(
      <ActionCellRenderer
        context={{ canDelete: true }}
        data={{ ident: mockIdent }}
        onDelete={mockOnDelete}
      />,
    );
    const elem = await screen.findByTestId('delete');
    fireEvent.click(elem);

    expect(mockOnDelete).toHaveBeenCalledTimes(1);
    expect(mockOnDelete).toHaveBeenCalledWith(mockIdent);
  });

  test('It shows the edit icon when allowed', async () => {
    render(
      <ActionCellRenderer
        context={{ canEdit: true }}
        data={{}}
      />,
    );

    expect(await screen.findByTestId('edit')).toBeInTheDocument();
  });

  test('onEdit is called when the edit icon is clicked', async () => {
    const mockOnEdit = jest.fn();
    const mockData = {};

    render(
      <ActionCellRenderer
        context={{ canEdit: true }}
        data={mockData}
        onEdit={mockOnEdit}
      />,
    );
    const elem = await screen.findByTestId('edit');
    fireEvent.click(elem);

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
    expect(mockOnEdit).toHaveBeenCalledWith(mockData);
  });

  test('It shows the fusion diagram icon when available', async () => {
    render(
      <ActionCellRenderer
        data={{ svg: mockSvg }}
      />,
    );

    expect(await screen.findByTestId('fusion')).toBeInTheDocument();
  });

  test('It shows the SVG dialog when the icon is clicked', async () => {
    render(
      <ActionCellRenderer
        data={{ svg: mockSvg }}
      />,
    );

    fireEvent.click(await screen.findByTestId('fusion'));
    expect(await screen.findByRole('presentation')).toBeInTheDocument();
  });

  test('It shows the image icon when available', async () => {
    render(
      <ActionCellRenderer
        data={{ image: mockImage }}
      />,
    );

    expect(await screen.findByTestId('image')).toBeInTheDocument();
  });

  test('It shows the image dialog when the icon is clicked', async () => {
    render(
      <ActionCellRenderer
        data={{ image: mockImage }}
      />,
    );

    fireEvent.click(await screen.findByTestId('image'));
    expect(await screen.findByRole('presentation')).toBeInTheDocument();
  });
});
