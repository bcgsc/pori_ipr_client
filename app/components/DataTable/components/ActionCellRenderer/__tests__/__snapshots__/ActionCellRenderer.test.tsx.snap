// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionCellRenderer It matches the snapshot 1`] = `<DocumentFragment />`;
